torch/testing/_internal/common_cuda.py

```diff
@@ -32,3 +35,41 @@ def initialize_cuda_context_rng():
         for i in range(torch.cuda.device_count()):
             torch.randn(1, device="cuda:{}".format(i))
         __cuda_ctx_rng_initialized = True
+
+
+def tf32_is_not_fp32():
+    if not torch.cuda.is_available():
+        return False
+    if torch.cuda.get_device_properties(torch.cuda.current_device()).major < 8:
+        return False
+    return True
+
+
+@contextlib.contextmanager
+def setup_tf32(dtype, rtol=0.001, atol=1e-5):
+    if dtype in {tfloat32, tcomplex64}:
+        torch.backends.cuda.matmul.allow_tf32 = True
+        if tf32_is_not_fp32():
+            yield tf32_to_fp32(dtype), rtol, atol
+        else:
+            yield tf32_to_fp32(dtype), None, None
+    else:
+        if dtype in {torch.float32, torch.complex64}:
+            torch.backends.cuda.matmul.allow_tf32 = False
+        yield dtype, None, None
+    torch.backends.cuda.matmul.allow_tf32 = True
+
+
+def tf32_on_and_off(rtol_=0.001, atol_=1e-5):
+    def wrapper(f):
+        @functools.wraps(f)
+        def wrapped(*args, **kwargs):
+            rtol, atol = (rtol_, atol_) if tf32_is_not_fp32() else (None, None)
+            torch.backends.cuda.matmul.allow_tf32 = True
```

Comment:
Instead of directly setting and resetting `torch.backends.cuda.matmul.allow_tf32` here, maybe you can use the context manager above, to ensure the `allow_tf32` global setting does not leak outside this test if an exception is thrown?

Reply:
`setup_tf32` requires a dtype argument, therefore not easily applicable. But I have modified the code to stash and recover the original flag.